### ISSUE TITLE
Referencing defined type throw error "unknown type"

### ISSUE THREAD

**Issue by Khalid-Nowaf** [hamba/avro#280]
A defined type can be used by referencing the name of the type.
```json
{
    "name": "a.b.rootType",
    "type": "record",
    "fields": [
        {
            "name": "someEnum1",
            "type": {
                "name": "a.b.rootType.classEnum",
                "type": "enum",
                "symbols": ["A", "B", "C"]
            }
        },
        {
            "name": "someEnum2",
            "type": "a.b.rootType.classEnum" // we have used the refance here
        }
    ]
}
```

This works as expected, but if we change the order such that 
```json
{
    "name": "a.b.rootType",
    "type": "record",
    "fields": [
       {
            "name": "someEnum2",
            "type": "a.b.rootType.classEnum"
        },
        {
            "name": "someEnum1",
            "type": {
                "name": "a.b.rootType.classEnum",
                "type": "enum",
                "symbols": ["A", "B", "C"]
            }
        }
    ]
}
```
The parser throws the following error: `Error: avro: unknown type: a.b.rootType.classEnum`

**Comment by Khalid-Nowaf** [comment]
I have looked at the code, and it seems it relies on the schema cache; if the type is not in the cache, the parser assumes the type definition is unknown. (and sometimes the type definition does exist, but it has not been cached yet)

I would see three approaches to fixing this issue:
1- a quick full schema scan cache the types
2- defer parsing any unknown type to the end, so the cache eventually will have the type definition if it does exist.
3- preprocess the schema by expanding any reference type so that each type would be populated with the exact definition.

**Comment by nrwiersma** [comment]
Hi,

I am reasonably certain the second case is not valid according to the spec. From my reading, name references must reference a previously defined name, meaning the definition for the name must be before the reference is used.

**Comment by Khalid-Nowaf** [comment]
@nrwiersma, thanks for a clear answer; you are correct; the definition must be defined first before using it, and now I just found the real issue in my code. Which is not related to ordering rather than naming. Here is an example of a schema that can be parsed in java `avro-tool`

```json
{
    "name": "a.b.rootType",
    "type": "record",
    "fields": [
        {
            "name": "someEnum1",
            "type": {
                "name": "classEnum",
                "type": "enum",
                "symbols": [
                    "A",
                    "B",
                    "C"
                ]
            }
        },
        {
            "name": "someEnum2",
            "type": "a.b.classEnum"
        }
    ]
}
```

but when I used it with `avrosv` I got the following error
```Error: avro: unknown type: a.b.classEnum```

**Comment by nrwiersma** [comment]
That does seem like a bug.

**Comment by nrwiersma** [comment]
Thanks for the contribution, it is released as part of [v2.14.1](https://github.com/hamba/avro/releases/tag/v2.14.1)
